src/main/scala/viper/gobra/reporting/StatsCollector.scala

```diff
@@ -0,0 +1,345 @@
+                             isTrusted: Boolean,
+                             isAbstract: Boolean,
+                             isImported: Boolean,
+                             isBuiltIn: Boolean)
+
+  override val name: String = "StatsCollector"
+
+  override def report(msg: GobraMessage): Unit = {
+    msg match {
+      // Capture typeInfo once it's available
+      case TypeInfoMessage(typeInfo, taskName) => this.synchronized({ typeInfos = typeInfos + (taskName -> typeInfo)})
+      // Free up unneeded space, once a task is finished
+      case VerificationTaskFinishedMessage(taskName) => this.synchronized({ typeInfos = typeInfos - taskName})
+      case GobraEntitySuccessMessage(taskName, _, e, info, time, cached) if !info.node.isInstanceOf[BuiltInMember] =>
+        Violation.violation(typeInfos.contains(taskName), "No type info available for stats reporter")
+        getMemberInformation(info.pnode, typeInfos(taskName)) match {
+          case GobraMemberInfo(pkgDir, pkg, memberName, args, isTrusted, isAbstract, isImported, false)  =>
+            addResult(
+              pkgDir,
+              pkg,
+              memberName,
+              info.pnode.getClass.getSimpleName,
+              args,
+              ViperMemberEntry(e, taskName, time, e.getClass.getSimpleName, success = true, cached = cached, isImported, hasBody = viperMemberHasBody(e)),
+              isTrusted,
+              isAbstract)
+          case _ =>
+        }
+      case GobraEntityFailureMessage(taskName, _, e, info, _, time, cached) if !info.node.isInstanceOf[BuiltInMember] =>
+        Violation.violation(typeInfos.contains(taskName), "No type info available for stats reporter")
+        getMemberInformation(info.pnode, typeInfos(taskName)) match {
+          case GobraMemberInfo(pkgDir, pkg, memberName, args, isTrusted, isAbstract, isImported, false) =>
+            addResult(
+              pkgDir,
+              pkg,
+              memberName,
+              info.pnode.getClass.getSimpleName,
+              args,
+              ViperMemberEntry(e, taskName, time, e.getClass.getSimpleName, success = false, cached = cached, isImported, hasBody = viperMemberHasBody(e)),
+              isTrusted,
+              isAbstract)
+          case _ =>
+        }
+      case _ =>
+    }
+    // Pass message to next reporter
+    reporter.report(msg)
+  }
+
+  def getNumberOfCachedViperMembers: Int = memberMap.values.flatMap(_.viperMembers.values).count(_.cached)
+
+  def getNumberOfCacheableViperMembers: Int = memberMap.values.flatMap(_.viperMembers.values).count(member => member.hasBody && member.nodeType == "Method")
+
+  def writeJsonReportToFile(file: File): Unit = {
+    if((file.exists() && file.canWrite) || file.getParentFile.canWrite) {
+      FileUtils.writeStringToFile(file, getJsonReport(false), UTF_8)
+    }
+  }
+
+  def getJsonReport(shorten: Boolean): String = {
+    val json = "[\n" +
+      memberMap.map({ case (key, gobraMember) => s"""  {
```

Comment:
I think that the code here is hard to read.

It would be better if we defined the "jsonification" of `memberMap`, `viperMembers`, and `dependencies` in separate variables and then introduced the result in the final string instead of doing this all in the declaration of variable `json`.

Reply:
Out of curiosity, would the classes `JSONArray` and `JSONObject` (http://lampwww.epfl.ch/~hmiller/scaladoc/library/scala/util/parsing/json/package.html) be useful here to avoid having to create this string "by hand"?